fabric-gametest-api-v1/src/main/java/net/fabricmc/fabric/impl/gametest/FabricGameTestModInitializer.java

```diff
@@ -16,49 +16,47 @@
 
 package net.fabricmc.fabric.impl.gametest;
 
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
-
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import net.minecraft.test.TestFunctions;
+import net.minecraft.registry.Registries;
+import net.minecraft.registry.Registry;
+import net.minecraft.registry.RegistryKeys;
+import net.minecraft.test.TestInstance;
+import net.minecraft.test.TestInstances;
 
 import net.fabricmc.api.ModInitializer;
+import net.fabricmc.fabric.api.event.registry.DynamicRegistrySetupCallback;
 import net.fabricmc.loader.api.FabricLoader;
-import net.fabricmc.loader.api.entrypoint.EntrypointContainer;
 
 public final class FabricGameTestModInitializer implements ModInitializer {
-	private static final String ENTRYPOINT_KEY = "fabric-gametest";
-	private static final Map<Class<?>, String> GAME_TEST_IDS = new HashMap<>();
 	private static final Logger LOGGER = LoggerFactory.getLogger(FabricGameTestModInitializer.class);
+	private static TestAnnotationLocator locator = new TestAnnotationLocator(FabricLoader.getInstance());
 
 	@Override
 	public void onInitialize() {
-		List<EntrypointContainer<Object>> entrypointContainers = FabricLoader.getInstance()
-				.getEntrypointContainers(ENTRYPOINT_KEY, Object.class);
-
-		for (EntrypointContainer<Object> container : entrypointContainers) {
-			Class<?> testClass = container.getEntrypoint().getClass();
-			String modid = container.getProvider().getMetadata().getId();
-
-			if (GAME_TEST_IDS.containsKey(testClass)) {
-				throw new UnsupportedOperationException("Test class (%s) has already been registered with mod (%s)".formatted(testClass.getCanonicalName(), modid));
-			}
-
-			GAME_TEST_IDS.put(testClass, modid);
-			TestFunctions.register(testClass);
-
-			LOGGER.debug("Registered test class {} for mod {}", testClass.getCanonicalName(), modid);
+		if (!(FabricGameTestRunner.ENABLED || FabricLoader.getInstance().isDevelopmentEnvironment())) {
+			// Don't try to load the tests if the game test runner is disabled or we are not in a development environment
+			return;
 		}
-	}
 
-	public static String getModIdForTestClass(Class<?> testClass) {
-		if (!GAME_TEST_IDS.containsKey(testClass)) {
-			throw new UnsupportedOperationException("The test class (%s) was not registered using the '%s' entrypoint".formatted(testClass.getCanonicalName(), ENTRYPOINT_KEY));
+		for (TestAnnotationLocator.TestMethod testMethod : locator.getTestMethods()) {
+			LOGGER.debug("Registering test method: {}", testMethod.identifier());
+			Registry.register(Registries.TEST_FUNCTION, testMethod.identifier(), testMethod.testFunction());
 		}
 
-		return GAME_TEST_IDS.get(testClass);
+		DynamicRegistrySetupCallback.EVENT.register(registryView -> {
+			// We ideally need an after loaded event, but this should hopefully work for now
```

Comment:
Yeah we do, this code is kinda horrible. If Mojang adds other builtin test instances, then the modded ones will be before them in the registry.

Reply:
Fixed, now uses a dedicated mixin. Id be somewhat hesitant to add an api for this into reg sync, I dont think we should encorage people to register dynmatic reg entries in code. Anyway thats a discussion for another time.